woocommerce/woocommerce-gateway-stripe
# Main settings page persistence

# Changes proposed in this Pull Request:

Fixes #1812.

This PR adds data API handling of all the settings on the WooCommerce > Settings > Payments > Stripe UPE page.

Clicking checkboxes or inputting values will update the `@wordpress/data` store, and clicking "Save changes" will submit a request to the back-end REST controller.

## A change in approach to settings store actions, hooks, and selectors

Since all of the selectors were very similar in function, we decided to simplify things by removing selectors and moving their logic into hooks. I also stole the idea to replace the actions with direct calls to `updateSettingsValues()` (introduced by @frosso in another PR). Since neither selectors or actions were no longer needed, all the logic (and future scope of changes) for retrieving and updating each setting is now limited to just `client/data/settings/hooks.js`.

I have also replaced the separate definitions for each hook with calls to a creator function - `makeSettingsHook()`.

I have also simplified tests by running the same set of tests for each hook.

Hopefully this will make it easier to add new hooks and discover bugs.

## REST controller test simplification

I have also simplified the REST controller tests (and added tests for untested fields) by extracting the shared test logic (similarly to the JS changes above) by field type (boolean, enum, or statement descriptor which has its own validator callback).

This allows for quickly adding new tests based on field type and making sure all the appropriate tests are run.

# Testing instructions

1. Enable UPE preview and UPE.
1. Go to the UPE settings page.
1. A loading placeholder should be shown while settings are loaded.
1. Change settings and save them.
1. Refresh the page - settings should be persisted and a snackbar notification confirming success should appear.
1. Disable UPE and confirm that the settings are presented with the same values in the old UI.
